### PR TITLE
OIDC Support

### LINKED ISSUE #234: support configuring openid-connect auth

### PR DESCRIPTION
**Pull request by kron4eg**
Fixes #234

```release-note
OIDC Support
```

### CODE REVIEW

File: config.yaml.dist [hunk 1]
```diff
@@ -69,7 +69,57 @@ features:
...
+  openid_connect:
+    enable: false
+
+    config:
+      # The URL of the OpenID issuer, only HTTPS scheme will be accepted. If
+      # set, it will be used to verify the OIDC JSON Web Token (JWT).
+      issuer_url: ""
+
+      # The client ID for the OpenID Connect client, must be set if
+      # oidc-issuer-url is set.
```
Comment (xmudrii) [inline config.yaml.dist]: Rename `oidc-issuer-url` to `issue_url`.

File: pkg/config/cluster.go [hunk 2]
```diff
@@ -355,6 +361,43 @@ type MetricsServer struct {
...
+
+// Validate features config
+func (f *Features) Validate() error {
+	// Currently only validate OIDC config
+	if !f.OpenIDConnect.Enable {
+		return nil
+	}
+
+	if f.OpenIDConnect.Config.IssuerURL == "" {
+		return errors.New("oidc-issuer-url can't be empty")
```
Comment (xmudrii) [inline pkg/config/cluster.go]: The error should mention the name of the field we use (`issuer_url`).

File: pkg/config/cluster.go [hunk 3]
```diff
@@ -355,6 +361,43 @@ type MetricsServer struct {
...
+	}
+
+	if f.OpenIDConnect.Config.ClientID == "" {
+		return errors.New("oidc-client-id can't be empty")
```
Comment (xmudrii) [inline pkg/config/cluster.go]: The error should mention the name of the field we use (`client_id`).

File: pkg/features/oidc.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-username-claim", feature.Config.UsernameClaim)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-username-prefix", feature.Config.UsernamePrefix)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-groups-claim", feature.Config.GroupsClaim)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-groups-prefix", feature.Config.GroupsPrefix)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-required-claim", feature.Config.RequiredClaim)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-signing-algs", feature.Config.SigningAlgs)
+	optinalMapSet(cfg.APIServer.ExtraArgs, "oidc-ca-file", feature.Config.CAFile)
+}
+
+func optinalMapSet(m map[string]string, key string, val string) {
```
Comment (xmudrii) [inline pkg/features/oidc.go]: What does `optinalMapSet` mean? Did you mean `optionalMapSet`?

File: pkg/config/cluster.go [hunk 5]
```diff
@@ -79,6 +80,10 @@ func (m *Cluster) DefaultAndValidate() error {
 		return errors.Wrap(err, "network configuration is invalid")
 	}
 
+	if err := m.Features.Validate(); err != nil {
+		return err
```
Comment (xmudrii) [inline pkg/config/cluster.go]: Usually, we expanded those error to include what validation failed, e.g. `oidc validation failed`.

**Comment by kron4eg** [comment]
/override pull-kubeone-e2e-digitalocean-conformance-1.13.5
/override pull-kubeone-e2e-digitalocean-conformance-1.14.1

DO API is dead for now

File: config.yaml.dist (hunk 1 above)
Comment (kron4eg) [reply config.yaml.dist]: ✔️

File: pkg/config/cluster.go (hunk 2 above)
Comment (kron4eg) [reply pkg/config/cluster.go]: ✔️

File: pkg/config/cluster.go (hunk 3 above)
Comment (kron4eg) [reply pkg/config/cluster.go]: ✔️

File: pkg/config/cluster.go (hunk 5 above)
Comment (kron4eg) [reply pkg/config/cluster.go]: ✔️

File: pkg/features/oidc.go (hunk 4 above)
Comment (kron4eg) [reply pkg/features/oidc.go]: typo 🤷‍♂️, fixed: ✔️

**Pull request merged**